### PR TITLE
[native] Add static mapping of native query configs

### PR DESCRIPTION
**Pull request by bikramSingh91**
This change updates the mapping of native only query configs from a
prefix-based system to a hardcoded, static mapping. This is to
decouple naming conventions that can be different between presto
and velox. For instance, Velox permits the use of dots, whereas Presto
employs dots to indicate the catalog to which a property should
be linked.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 1]
```diff
@@ -311,7 +311,7 @@ public final class SystemSessionProperties
     public static final String NATIVE_EXECUTION_EXECUTABLE_PATH = "native_execution_executable_path";
     public static final String NATIVE_EXECUTION_PROGRAM_ARGUMENTS = "native_execution_program_arguments";
     public static final String NATIVE_EXECUTION_PROCESS_REUSE_ENABLED = "native_execution_process_reuse_enabled";
-    public static final String NATIVE_DEBUG_VALIDATE_OUTPUT_FROM_OPERATORS = "native_debug.validate_output_from_operators";
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:314]: Would you extract this change into a separate commit?

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp [hunk 2]
```diff
@@ -51,14 +52,30 @@ folly::IOThreadPoolExecutor* spillExecutorPtr() {
...
 namespace {
-std::string maybeRemoveNativePrefix(const std::string& name) {
-  static const std::string kNativePrefix = "native_";
-  const auto result =
-      ::strncmp(name.c_str(), kNativePrefix.c_str(), kNativePrefix.size());
-  if (result == 0) {
-    return name.substr(kNativePrefix.length());
-  }
-  return name;
+std::string fromPrestoToNativeConfig(const std::string& name) {
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/QueryContextManager.cpp:55]: nit: toVeloxConfig

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp [hunk 3]
```diff
@@ -51,14 +52,30 @@ folly::IOThreadPoolExecutor* spillExecutorPtr() {
...
+  using velox::core::QueryConfig;
+  static const std::unordered_map<std::string, std::string>
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/QueryContextManager.cpp:57]: F14FastMap?

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp [hunk 4]
```diff
@@ -51,14 +52,30 @@ folly::IOThreadPoolExecutor* spillExecutorPtr() {
...
+      MAP_FROM_PRESTO_TO_NATIVE_CONFIG = {
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/QueryContextManager.cpp:58]: kPrestoToVeloxMapping

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp [hunk 5]
```diff
@@ -51,14 +52,30 @@ folly::IOThreadPoolExecutor* spillExecutorPtr() {
...
+          {"native_spill_compression_codec",
+           QueryConfig::kSpillCompressionKind},
+          {"native_spill_write_buffer_size",
+           QueryConfig::kSpillWriteBufferSize},
+          {"native_join_spill_enabled", QueryConfig::kJoinSpillEnabled},
+          {"native_execution_process_reuse_enabled",
+           QueryConfig::kAggregationSpillMemoryThreshold},
+          {"native_debug_validate_output_from_operators",
+           QueryConfig::kValidateOutputFromOperators}};
+  auto entry = MAP_FROM_PRESTO_TO_NATIVE_CONFIG.find(name);
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/QueryContextManager.cpp:77]: entry -> it

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp (hunk 2 above)
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/QueryContextManager.cpp:55]: Would you add a comment to explain how this function works, e.g. clarify that it returns 'name' as is if there is no mapping.

**Review by mbasmanova — COMMENTED** [review]
Thanks. Looks good % a few minor comments.

**Comment by lingbin** [comment]
nit but as a reminder: In my observation, in the title of the commit message, it is more likely to use `"[native]"`(lowercase) as a tag instead of `"[Native]"`.

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp (hunk 2 above)
Comment (kewang1024) [reply presto-native-execution/presto_cpp/main/QueryContextManager.cpp:55]: +1

**Review by kewang1024 — COMMENTED** [review]
RE "For example velox allows dots whereas presto does not."

Can you give an example of such config/session property?

**Comment by bikramSingh91** [comment]
> RE "For example velox allows dots whereas presto does not."
> 
> Can you give an example of such config/session property?

@kewang1024  `debug.validate_output_from_operators` is one such config that uses dots whereas in Presto, it employs dots to indicate the catalog to which a property should be linked. https://prestodb.io/docs/current/sql/set-session.html

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp [hunk 6]
```diff
@@ -51,14 +52,32 @@ folly::IOThreadPoolExecutor* spillExecutorPtr() {
...
-  if (result == 0) {
-    return name.substr(kNativePrefix.length());
-  }
-  return name;
+// Utility function to translate a config name in presto to its equivalent in
+// Velox. Returns 'name' as is if there is no mapping.
+std::string toVeloxConfig(const std::string& name) {
+  using velox::core::QueryConfig;
+  static const folly::F14FastMap<std::string, std::string>
+      kPrestoToVeloxMapping = {
```
[the remark this answers is not in the document]
Comment (bikramSingh91) [reply presto-native-execution/presto_cpp/main/QueryContextManager.cpp:60]: The following properties do not map to anything in velox:
NATIVE_SIMPLIFIED_EXPRESSION_EVALUATION_ENABLED
NATIVE_EXECUTION_ENABLED
NATIVE_EXECUTION_EXECUTABLE_PATH
NATIVE_EXECUTION_PROGRAM_ARGUMENTS
NATIVE_EXECUTION_PROCESS_REUSE_ENABLED

I looked again and found "expression.eval_simplified" that looks related to native_simplified_expression_evaluation_enabled and will add it in the next iteration of this PR. This however means native_simplified_expression_evaluation_enabled actually never worked as there is no "simplified_expression_evaluation_enabled" (after removing `native_` prefix) config in velox.

The rest are used inside presto code, so dont need any explicit mapping in velox

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp [hunk 7]
```diff
@@ -51,14 +52,32 @@ folly::IOThreadPoolExecutor* spillExecutorPtr() {
...
+           QueryConfig::kOrderBySpillMemoryThreshold},
+          {"native_max_spill_level", QueryConfig::kMaxSpillLevel},
+          {"native_max_spill_file_size", QueryConfig::kMaxSpillFileSize},
+          {"native_spill_compression_codec",
+           QueryConfig::kSpillCompressionKind},
+          {"native_spill_write_buffer_size",
+           QueryConfig::kSpillWriteBufferSize},
+          {"native_join_spill_enabled", QueryConfig::kJoinSpillEnabled},
+          {"native_execution_process_reuse_enabled",
+           QueryConfig::kAggregationSpillMemoryThreshold},
```
[the remark this answers is not in the document]
Comment (bikramSingh91) [reply presto-native-execution/presto_cpp/main/QueryContextManager.cpp:76]: Thanks for catching this, forgot to remove it since it doesnt map to anything in velox. Will remove it in the next iteration of this PR

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (kewang1024) [reply presto-native-execution/presto_cpp/main/QueryContextManager.cpp:60]: I see, thanks for checking!

**Review by kewang1024 — APPROVED** [review]
LGTM, thanks!

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp [hunk 8]
```diff
@@ -51,14 +52,32 @@ folly::IOThreadPoolExecutor* spillExecutorPtr() {
...
 namespace {
-std::string maybeRemoveNativePrefix(const std::string& name) {
-  static const std::string kNativePrefix = "native_";
-  const auto result =
-      ::strncmp(name.c_str(), kNativePrefix.c_str(), kNativePrefix.size());
-  if (result == 0) {
-    return name.substr(kNativePrefix.length());
-  }
-  return name;
+// Utility function to translate a config name in presto to its equivalent in
```
Comment (mbasmanova) [inline presto-native-execution/presto_cpp/main/QueryContextManager.cpp:55]: nit: presto -> Presto

**Review by mbasmanova — APPROVED** [review]
Thanks.

**Comment by bikramSingh91** [comment]
A failure in testQueryScanExceeded in pre-commits which looks unrelated

**Review by xiaoxmeng — APPROVED** [review]
@bikramSingh91 LGTM. Thanks!
